subscriber_notifee.go

```diff
@@ -14,82 +16,201 @@ import (
+			return nil, fmt.Errorf("dht not subscribed to local routability events; err: %s", err)
+		}
+	}
+
+	nn := &subscriberNotifee{
+		dht:            dht,
+		identifySub:    identifySub,
+		updateSub:      updateSub,
+		routabilitySub: routabilitySub,
 	}
-	defer sub.Close()
 
+	// register for network notifications
+	dht.host.Network().Notify(nn)
+
+	// Fill routing table with currently connected peers that are DHT servers
 	dht.plk.Lock()
+	defer dht.plk.Unlock()
 	for _, p := range dht.host.Network().Peers() {
 		protos, err := dht.peerstore.SupportsProtocols(p, dht.protocolStrs()...)
-		if err == nil && len(protos) != 0 {
+		if err != nil {
+			return nil, fmt.Errorf("could not check peerstore for protocol support: err: %s", err)
+		}
+		if len(protos) != 0 {
 			dht.Update(dht.ctx, p)
 		}
 	}
-	dht.plk.Unlock()
+
+	return nn, nil
+}
+
+func (nn *subscriberNotifee) subscribe(proc goprocess.Process) {
+	dht := nn.dht
+	defer dht.host.Network().StopNotify(nn)
+	defer nn.identifySub.Close()
+	defer nn.updateSub.Close()
+
+	// skip routability events if they are not enabled
+	var routabilityEvts <-chan interface{}
+	if nn.routabilitySub != nil {
+		defer nn.routabilitySub.Close()
+		routabilityEvts = nn.routabilitySub.Out()
+	}
 
 	for {
 		select {
-		case evt, more := <-sub.Out():
-			// we will not be getting any more events
+		case evt, more := <-nn.identifySub.Out():
 			if !more {
 				return
 			}
-
-			// something has gone really wrong if we get an event for another type
-			ev, ok := evt.(event.EvtPeerIdentificationCompleted)
-			if !ok {
-				logger.Errorf("got wrong type from subscription: %T", ev)
+			handlePeerIdentificationCompletedEvent(dht, evt)
+		case evt, more := <-nn.updateSub.Out():
+			if !more {
 				return
 			}
-
-			dht.plk.Lock()
-			if dht.host.Network().Connectedness(ev.Peer) != network.Connected {
-				dht.plk.Unlock()
-				continue
-			}
-
-			// if the peer supports the DHT protocol, add it to our RT and kick a refresh if needed
-			protos, err := dht.peerstore.SupportsProtocols(ev.Peer, dht.protocolStrs()...)
-			if err == nil && len(protos) != 0 {
-				refresh := dht.routingTable.Size() <= minRTRefreshThreshold
-				dht.Update(dht.ctx, ev.Peer)
-				if refresh && dht.autoRefresh {
-					select {
-					case dht.triggerRtRefresh <- nil:
-					default:
-					}
-				}
+			handlePeerProtocolsUpdatedEvent(dht, evt)
+		case evt, more := <-routabilityEvts:
+			if !more {
+				return
 			}
-			dht.plk.Unlock()
-
+			handleRoutabilityChanges(dht, evt)
 		case <-proc.Closing():
 			return
 		}
 	}
 }
 
+func handlePeerIdentificationCompletedEvent(dht *IpfsDHT, evt interface{}) {
+	// something has gone really wrong if we get an event for another type
+	e, ok := evt.(event.EvtPeerIdentificationCompleted)
+	if !ok {
+		logger.Errorf("got wrong type from subscription: %T", e)
+		return
+	}
+
+	dht.plk.Lock()
+	defer dht.plk.Unlock()
+	if dht.host.Network().Connectedness(e.Peer) != network.Connected {
+		return
+	}
```

Comment:
Note: I believe both this check and this lock can go away once we switch to the new routing table logic. At that point, it's actually perfectly fine to add disconnected peers to our routing table.

Reply:
@aarshkshah1992 can you confirm, I haven't had time to look through the routing table PR yet.

Reply:
I talked with him about this and I was wrong. We _can_ make this possible with a small tweak to the routing table, but we'll handle that later.

(small tweak == get rid of live/dead states, just call a "checkPeer" function on all peers every tick)